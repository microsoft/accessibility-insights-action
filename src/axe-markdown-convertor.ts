// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AxeScanResults } from 'accessibility-insights-scan';
import { stripIndent } from 'common-tags';
import { injectable } from 'inversify';
import { isEmpty } from 'lodash';
import * as table from 'markdown-table';

import { toolName } from './content/strings';

@injectable()
export class AxeMarkdownConvertor {
    public convert(axeScanResults: AxeScanResults): string {
        if (isEmpty(axeScanResults.results.violations)) {
            return this.getCongratsText();
        }

        return this.getFailureDetails(axeScanResults);
    }

    private getFailureDetails(axeScanResults: AxeScanResults): string {
        return stripIndent`
        FAILED RULES:

${this.getFailedRuleTable(axeScanResults)}
        `;
    }

    private getCongratsText(): string {
        return stripIndent`
            Congratulations!
            No failed automated checks were found.
        `;
    }

    private getFailedRuleTable(axeScanResults: AxeScanResults): string {
        const tableContent: string[][] = [['Rule', 'Count']];

        const violations = axeScanResults.results.violations;

        violations.forEach(violation => {
            const row: string[] = [violation.id, violation.nodes.length.toString()];
            tableContent.push(row);
        });

        return table(tableContent);
    }
}
