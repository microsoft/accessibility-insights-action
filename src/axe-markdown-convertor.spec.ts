// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
// tslint:disable:no-import-side-effect no-any
import 'reflect-metadata';

import { AxeScanResults } from 'accessibility-insights-scan';

import { AxeMarkdownConvertor } from './axe-markdown-convertor';

// tslint:disable: no-unsafe-any no-null-keyword no-object-literal-type-assertion
describe(AxeMarkdownConvertor, () => {
    let axeMarkdownConvertor: AxeMarkdownConvertor;
    let axeScanResults: AxeScanResults;

    beforeEach(() => {
        axeMarkdownConvertor = new AxeMarkdownConvertor();
        axeScanResults = {
            results: {
                violations: [],
            },
        } as AxeScanResults;
    });

    it('should create instance', () => {
        expect(axeMarkdownConvertor).not.toBeNull();
    });

    describe('convert', () => {
        it('returns congrats message when no failure found', async () => {
            const res = axeMarkdownConvertor.convert(axeScanResults);

            expect(res).toMatchSnapshot();
        });

        it('returns failure details', async () => {
            axeScanResults.results = {
                violations: [
                    {
                        id: 'color-contrast',
                        nodes: [{ html: 'html1' }, { html: 'html2' }],
                    },
                    {
                        id: 'duplicate-id',
                        nodes: [{ html: 'html3' }, { html: 'html4' }, { html: 'html5' }],
                    },
                ],
                passes: [{ html: 'passed' }],
                inapplicable: [{ html: 'inapplicable' }],
            } as any;

            const res = axeMarkdownConvertor.convert(axeScanResults);

            expect(res).toMatchSnapshot();
        });
    });
});
