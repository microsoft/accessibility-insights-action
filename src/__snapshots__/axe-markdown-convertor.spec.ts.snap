// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AxeMarkdownConvertor convert returns congrats message when no failure found 1`] = `
"Congratulations!
No failed automated checks were found."
`;

exports[`AxeMarkdownConvertor convert returns failure details 1`] = `
"FAILED RULES:

| Rule           | Count |
| -------------- | ----- |
| color-contrast | 2     |
| duplicate-id   | 3     |"
`;
